### ISSUE TITLE
Elements from this relationship are not deleted when an empty array is set

### ISSUE THREAD

**Issue by Flexiink** [typeorm/typeorm#2121]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

```
class ProductsModel extends ORM {
    @OneToMany(type => ProductProductPropertiesModel, prodProdProperies => prodProdProperies.product, {
        eager: true,
        cascade: true,
        onDelete: 'CASCADE',
    })
    @JoinTable({
        name: 'product-product-properties',
        joinColumn: {
            name: "product_id",
            referencedColumnName: "id"
        },
        inverseJoinColumn: {
            name: "propuct_property_id",
            referencedColumnName: "id"
        }
    })
    productProductProperties: ProductProductPropertiesModel[];
}

export class ProductProductPropertiesModel extends ORM {
    @ManyToOne(type => ProductsModel, product => product.id, { nullable: false })
    @JoinColumn({ name: 'product_id' })
    product: ProductsModel;}

-------------------------

product.productProductProperties = [];
product.save();

-------------------------
Unhandled Promise rejection: ER_BAD_NULL_ERROR: Column 'product_id' cannot be null ; Zone: <root> ; Task: Promise.then ; Value: { QueryFailedError: ER_BAD_NULL_ERROR: Column 'product_id' cannot be null

```
How to automatically delete these records?

**Comment by Flexiink** [comment]
I made your recommendations
```
export class ProductsModel extends ORM {
@OneToMany(type => ProductProductPropertiesModel, prodProdProperies => prodProdProperies.product, { eager: true, cascade: true })
    productProductProperties: ProductProductPropertiesModel[];
}

export class ProductProductPropertiesModel extends ORM {
    @ManyToOne(type => ProductsModel, product => product.id, { nullable: false, onDelete: 'CASCADE', })
    @JoinColumn({ name: 'product_id' })
    product: ProductsModel;
}
```
but I still get this error
```
Unhandled Promise rejection: ER_BAD_NULL_ERROR: Column 'product_id' cannot be null ; Zone: <root> ; Task: Promise.then ; Value: { QueryFailedError: ER_BAD_NULL_ERROR: Column 'product_id' cannot be null
```
I have to manually delete.
```
await getRepository(ProductProductPropertiesModel).delete(property.id);
```
it is right?

**Comment by pleerock** [comment]
So you have an error:

> Column 'product_id' cannot be null

try to figure out why it says so

**Comment by Kononnable** [comment]
@Flexiink 
Did you resolve your issue?

**Comment by Kononnable** [comment]
Closing as no answer from the reporter.

**Comment by kelvin-evari** [comment]
Sorry I dig up this one but I'm having the same problem.
> So you have an error:
> 
> > Column 'product_id' cannot be null
> 
> try to figure out why it says so

This is because whenever updating an array, typeorm compares the new and old array to find which records should be remove. However, it set the value of foreign key to be null instead of removing those records:
`UPDATE "ProductProductPropertiesModel" SET "product_id" = $2 WHERE "id" = $1 -- PARAMETERS: [<id of should-be-deleted record>,null]`.
Since this column is set to `nullable: false`, this query causes the issue.

I have to remove those old records manually, which I don't think is a good approach. Not sure why typeorm doesn't remove them instead?

**Comment by naisk133** [comment]
Sorry but, any update on this issue?

TypeORM version 0.2.18 still have issue like @kelvin-evari said that instead of deleting the relation record, it update the record and set the column id to null.

**Comment by VRspace4** [comment]
Having the same issue. Manually checking and deleting the records is going to be challenging since I have a large number of relationships.

Please reopen.

**Comment by mfaizanse** [comment]
I am facing the same issue even when using:
```
await Community.createQueryBuilder()
      .relation(Community, "members")
      .of(id)
      .remove(memberId);
```

Error:
```
query failed: UPDATE "community_member" SET "community_id" = $3 WHERE "community_id" = $1 AND "person_id" = $2 -- PARAMETERS: [5,1,null]
error: { error: null value in column "community_id" violates not-null constraint
```

**Comment by soylomass** [comment]
Have the same issue. Typeorm does not delete, only sets the id to NULL / 0

**Comment by xxRockOnxx** [comment]
+1 still existing.

**Comment by Migiyaka** [comment]
Similar issue, I have entities "questionnaire" and "question", and a unique composite key on the "question" entity consisting of "questionnaireId" + "order". When I try to pass an entirely new array of questions inside an update method for questionnaire, I get the error "Duplicate entry for key..." because I'm guessing the save method first tries to create new entries and only after update the old ones (and not even delete them for some reason).

Do I really have to do this manually?

**Comment by matvi3nko** [comment]
On save entity with One-to-Many relations:

It inserts new relations and then inserts NULL for old, but if not all entries are absolutely new, and there are intersections, TypeORM will change everything as NULL.

`query: START TRANSACTION
query: INSERT INTO `User_UserRoles`(`UserName`, `UserRoleId`) VALUES (?, ?) -- PARAMETERS: ["first-user@mail.com",4]
query: UPDATE `User_UserRoles` SET `UserName` = ? WHERE `UserName` = ? AND `UserRoleId` = ? -- PARAMETERS: [null,"first-user@mail.com",4]
query: COMMIT`

Try with twice save/update the same entity.

**Comment by luchfilip** [comment]
Same here. Can't create a new entity, keeps throwing this error:
```
message:"ER_BAD_NULL_ERROR: Column 'id' cannot be null"
sql:"INSERT INTO `user_card`(`id`, `createdAt`, `deletedAt`, `firstname`, `lastname`, `fullname`, `profileTitle`, `introDescription`, `photo`, `url`, `qrCode`, `userId`) VALUES (NULL, DEFAULT, DEFAULT, 'Filz', NULL, 'Filz', NULL, NULL, NULL, NULL, NULL, '7e9b75f5-d678-425a-acd7-bc68307bcf10')"
sqlMessage:"Column 'id' cannot be null"
sqlState:"23000"
```

**Comment by nickhaughton** [comment]
+1. Shouldn't removed elements respect the `onDelete` property? Right now the field is set to null, even if the field is marked as not nullable.

**Comment by Piranit** [comment]
same here
I have lazy relation and want set to empty array before inserting new entities
`manager.accessibleSector = Promise.resolve([]);`
`await this.managerRepository.save(manager);`
and received
`error: Error: ER_BAD_NULL_ERROR: Column 'managerUUID' cannot be null`

**Comment by itsmamipiyur** [comment]
This is still an existing problem.

**Comment by alonsosfera** [comment]
Any solution to this?

**Comment by adrianaxente** [comment]
Any updates I am facing the same issue and this is quite annoying!

**Comment by jaequery** [comment]
Hm, having worked with several others ORMs, I am to be quite frank, surprised that this type of issue exists and for this long on TypeORM. This is kinda like the basic features of what most ORMs provide when handling relationships.

**Comment by CentreleARES** [comment]
Any workarounds?

**Comment by martinsantibanez** [comment]
@pleerock any updates, or workarounds?

**Comment by martinsantibanez** [comment]
@pleerock any update?

**Comment by nebkat** [comment]
I believe this may be solved with #7105?

---

There is really no need for 7 comments in a row checking whether anything has changed here. It is unfortunate that this has not yet been resolved but it will not happen any faster with more comments. Unless you have something useful to add that will help in solving the problem, use the :+1: button on the existing discussions.

**Comment by i5o** [comment]
In case you reached here through Google, the solution that @nebkat provides does work.
![image](https://user-images.githubusercontent.com/2650479/109538744-36a36a00-7a9f-11eb-9513-7a3151936b02.png)
